Adding ML information (#2)
RedHatQuickCourses/rhods-model: modules/chapter1/pages/section1.adoc

. It is difficult or impossible to explicitly describe the solution or criteria behind a decision.
. It is not possible to develop an optimal solution by using classic programming paradigms, such as imperative programming.
. You have labeled data, or existing examples where you can describe the situation and map it to the correct result.
. You are looking for new areas of research and/or insights in large data sets.

== Types of Machine Learning

Supervised learning::

> Learning from labeled data, meaning the variables and categories are already known. The model is instructed to learn from the training data andmake predictions on unseen data. E.g estimate the price of a house, recognize which animal is present in a picture. Examples of algorithms fitting into this category are linear regressions, logistic regressions, support vector machines (SVMs), neural networks (deep or simple), decision trees, random forests and others. 

Reviewer: typos 

```suggestion
Learning from labeled data, meaning the variables and categories are already known. The model is instructed to learn from the training data and make predictions on unseen data. E.g. estimate the price of a house, recognize which animal is present in a picture. Examples of algorithms fitting into this category are linear regressions, logistic regressions, support vector machines (SVMs), neural networks (deep or simple), decision trees, random forests and others. 
```

Unsupervised learning:: 

> Learning from unlabeled data to discover patterns or relationships without explicit guidance as to what exists in the data. The model discovers hidden structures or insights within the data. E.g Identify different segments or groups of customers based on their spending habits, items that have a high possibility of being purchased together in a supermarket, anomaly detection. Examples of algorithms fitting into this category are K-means clustering, K-nearest neighbords(KNN), Density-based spatial clusering of applications with noise (DBSCAN), Principal Component Analysis(PCA) and autoencoders. 

Reviewer: ```suggestion
Learning from unlabeled data to discover patterns or relationships without explicit guidance as to what exists in the data. The model discovers hidden structures or insights within the data. E.g. identify different segments or groups of customers based on their spending habits, items that have a high possibility of being purchased together in a supermarket, anomaly detection. Examples of algorithms fitting into this category are K-means clustering, K-nearest neighbords(KNN), Density-based spatial clusering of applications with noise (DBSCAN), Principal Component Analysis(PCA) and autoencoders. 
```

Semi-Supervised Learning:: 

> Semi-Supervised learning is a machine learning algorithm that works between the supervised and unsupervised learning. It uses a small portion of  labelled and a large majority of unlabelled data. The labelled data is used to provide order to the learning problem while the unlabeled data is used to create a more generalized model by providing context. The upside is we are able to create more accurate and resilliant models by negating the caveats that exists for supervised and unsupervised learning.  It is particularly useful when obtaining labelled data is costly, time-consuming, or resource-intensive. One example of application are is Image and Speech analysis. 

Reviewer: ```suggestion
Semi-Supervised learning is a machine learning algorithm that works between the supervised and unsupervised learning. It uses a small portion of  labelled and a large majority of unlabeled data. The labelled data is used to provide order to the learning problem while the unlabeled data is used to create a more generalized model by providing context. The upside is we are able to create more accurate and resilient models by negating the caveats that exists for supervised and unsupervised learning.  It is particularly useful when obtaining labelled data is costly, time-consuming, or resource-intensive. Examples of semi-supervised learning are image and speech analysis. 
```

Reinforcement learning:: 

> In comparison to the previous groups of algorithms mentioned, reinforcement learning does not have the goal of grouping or classifying data. It tries to find the most optimal path. It involves an agent learning to make decisions through trial and error interactions with an environment. The agent receives feedback in the form of rewards or penalties for its actions, guiding it towards achieving a goal by maximizing the rewards. E.g Autonomous driving, AlphaGo learning to play Go by playing against itself. Examples of algorithms are Monte Carlo, Deep Q Networks(DQN), State-Action-Reward-State-Acction(SARSA) and others. 

Reviewer: ```suggestion
In comparison to the previous groups of algorithms mentioned, reinforcement learning does not have the goal of grouping or classifying data. It tries to find the most optimal path. It involves an agent learning to make decisions through trial and error interactions with an environment. The agent receives feedback in the form of rewards or penalties for its actions, guiding it towards achieving a goal by maximizing the rewards. E.g. Autonomous driving, AlphaGo learning to play Go by playing against itself. Examples of algorithms are Monte Carlo, Deep Q Networks(DQN), State-Action-Reward-State-Acction(SARSA) and others. 
```

== Choosing your algorithm
We have a plethora of mathematical algorithms that may, or may not, solve our problem and be applied on our data. But, how do you recognise the category of problem and the algorithm that can help you solve it?
The clue lies with your outcome variable.
 

> In a classification problem, the outcome variable (the event you want to predict) is a category or class label. It can be binary, such as "Sick" or "Healthy", "Yes" or "No". They can be multi-class such as for image recognition when you want to identify more than 2 classes or labels. 

Reviewer: Expanding the example a bit and adding a mention to multi-label problems.

```suggestion
In a classification problem, the outcome variable (the event you want to predict) is a category or class label. It can be binary, such as "Sick" or "Healthy", "Yes" or "No". The problem can be multi-class, such as for image recognition, when you want to identify more than two classes or labels, for example, dog breeds. Lastly, you can also have multi-label problems, where the model assigns more than one label to the same record. For example, a model might detect multiple labels in a single  music clip, e.g. rock, pop, 90s.
```

In a regression problem, the outcome variable is a continuous value and the goal is predicting a quantitative value. Common examples are predicting house prices, temperatures and sale revenues. 

Knowing the outcome variable you are able to identify if its supervised, unsupervised, semi-supervised or reinforcement learning. However, there is a lot of work left before the final algorithm is chosen. The data scientist will spend time exploring the data set, applying different visualization and data cleaning technqiues exploring the underlying data, and several different machine learning algorithms may be applied to test which one solves the problem. Solving the problem is multifacetted and there is no one algorithm to rule them all.  
 
== What is a Machine Learning model

> A Machine Learning model is a file that has been trained to recognize certain types of patterns. You train a mdel over a set of data, providing it with an algorithm that it can use to reason over and learn from that data.

Reviewer: ```suggestion
A Machine Learning model is a file that has been trained to recognize certain types of patterns. You train a model over a set of data, providing it with an algorithm that it can use to reason over and learn from that data.
```

> Once you have trained the model, you can use it to reason over data that it has not seen before but comes from the same applicability domain, and make predictions about that data. For example, let's say you want to build an application that can recognize a user's emotions based on their facial expressions. You can train a model by providing it with images of faces that are each tagged with a certain emotion, and then you can use that model in an application that can recognize any user's emotion.

Reviewer: What if we briefly mention the "generalization" concept?

```suggestion
Once you have trained the model, you can use it to reason over data that it has not seen before but comes from the same applicability domain, and make predictions about that data. This ability, called _generalization_, is a key concept in machine learning, as it refers to the capacity of models to adapt to unseen data.

 For example, let's say you want to build an application that can recognize a user's emotions based on their facial expressions. You can train a model by providing it with images of faces that are each tagged with a certain emotion, and then you can use that model in an application that can recognize any user's emotion.
```

Author: Love it!

. It is difficult or impossible to explicitly describe the solution or criteria behind a decision.
. It is not possible to develop an optimal solution by using classic programming paradigms, such as imperative programming.
. You have labeled data, or existing examples where you can describe the situation and map it to the correct result.
. You are looking for new areas of research and/or insights in large data sets.

== Types of Machine Learning

Supervised learning::

> Learning from labeled data, meaning the variables and categories are already known. The model is instructed to learn from the training data and make predictions on unseen data. E.g. estimate the price of a house, recognize which animal is present in a picture. Examples of algorithms fitting into this category are linear regressions, logistic regressions, support vector machines (SVMs), neural networks (deep or simple), decision trees, random forests and others. 

Reviewer 2: ```suggestion
Learning from labeled data, meaning the variables and categories are already known. The model is instructed to learn from the training data and make predictions on unseen data, for instance when estimating the price of a house or recognizing which animal is present in a picture. Examples of algorithms fitting into this category are linear regressions, logistic regressions, support vector machines (SVMs), neural networks (deep or simple), decision trees, and random forests.
```

Unsupervised learning:: 

> Learning from unlabeled data to discover patterns or relationships without explicit guidance as to what exists in the data. The model discovers hidden structures or insights within the data. E.g. identify different segments or groups of customers based on their spending habits, items that have a high possibility of being purchased together in a supermarket, anomaly detection. Examples of algorithms fitting into this category are K-means clustering, K-nearest neighbords(KNN), Density-based spatial clusering of applications with noise (DBSCAN), Principal Component Analysis(PCA) and autoencoders. 

Reviewer 2: ```suggestion
Learning from unlabeled data to discover patterns or relationships without explicit guidance as to what exists in the data. The model discovers hidden structures or insights within the data, for instance to identify different segments or groups of customers based on their spending habits, items that have a high possibility of being purchased together in a supermarket, or anomalous behavior. Examples of algorithms fitting into this category are K-means clustering, K-nearest neighbords (KNN), density-based spatial clustering of applications with noise (DBSCAN), Principal Component Analysis (PCA) and autoencoders. 
```

Semi-Supervised Learning:: 

> Semi-Supervised learning is a machine learning algorithm that works between the supervised and unsupervised learning. It uses a small portion of  labelled and a large majority of unlabeled data. The labelled data is used to provide order to the learning problem while the unlabeled data is used to create a more generalized model by providing context. The upside is we are able to create more accurate and resilient models by negating the caveats that exists for supervised and unsupervised learning.  It is particularly useful when obtaining labelled data is costly, time-consuming, or resource-intensive. Examples of semi-supervised learning are image and speech analysis. 

Reviewer 2: ```suggestion
Semi-Supervised learning is a machine learning algorithm that works between the supervised and unsupervised learning. It uses a small portion of labelled and a large majority of unlabeled data. The labelled data is used to provide order to the learning problem while the unlabeled data is used to create a more generalized model by providing context. It allows us to create more accurate and resilient models by negating the caveats that exists for supervised and unsupervised learning.  It is particularly useful when obtaining labelled data is costly, time-consuming, or resource-intensive. Semi-supervised learning is applied in image and speech analysis, for example. 
```

Reinforcement learning:: 

> In comparison to the previous groups of algorithms mentioned, reinforcement learning does not have the goal of grouping or classifying data. It tries to find the most optimal path. It involves an agent learning to make decisions through trial and error interactions with an environment. The agent receives feedback in the form of rewards or penalties for its actions, guiding it towards achieving a goal by maximizing the rewards. E.g. Autonomous driving, AlphaGo learning to play Go by playing against itself. Examples of algorithms are Monte Carlo, Deep Q Networks(DQN), State-Action-Reward-State-Acction(SARSA) and others. 

Reviewer 2: ```suggestion
In comparison to the previous groups of algorithms mentioned, reinforcement learning does not have the goal of grouping or classifying data. It is applied to find the optimal behavior of an agent for a given learning task. The agent learns to make decisions through trial and error interactions with an environment, and it receives feedback in the form of rewards or penalties for its actions, guiding it towards achieving a goal by maximizing the rewards. Noteworthy examples are autonomous driving or AlphaGo learning to play Go by playing against itself. Typical reinforcement learning algorithms are Monte Carlo, Deep Q Networks(DQN), State-Action-Reward-State-Acction(SARSA) and others. 
```

Knowing the outcome variable you are able to identify if its supervised, unsupervised, semi-supervised or reinforcement learning. However, there is a lot of work left before the final algorithm is chosen. The data scientist will spend time exploring the data set, applying different visualization and data cleaning technqiues exploring the underlying data, and several different machine learning algorithms may be applied to test which one solves the problem. Solving the problem is multifacetted and there is no one algorithm to rule them all.  
 
== What is a Machine Learning model

A Machine Learning model is a file that has been trained to recognize certain types of patterns. You train a model over a set of data, providing it with an algorithm that it can use to reason over and learn from that data.

Once you have trained the model, you can use it to reason over data that it has not seen before but comes from the same applicability domain, and make predictions about that data. This ability, called _generalization_, is a key concept in machine learning, as it refers to the capacity of models to adapt to unseen data.

>  For example, let's say you want to build an application that can recognize a user's emotions based on their facial expressions. You can train a model by providing it with images of faces that are each tagged with a certain emotion, and then you can use that model in an application that can recognize any user's emotion.

Reviewer 2: ```suggestion
For example, let's say you want to build an application that can recognize a user's emotions based on their facial expressions. You can train a model by providing it with images of faces that are each tagged with a certain emotion, and then you can use that model in an application that can recognize any user's emotion.
```
